File: day06/tuning_trouble.py
from sys import argv

def marker_pos(buf):
    for i in range(len(buf)-4):
        if len(set(buf[i:i+4])) == 4:
            return i+4
    return -1

with open(argv[1]) as f:
    for line in f:
        print('Part 1:', marker_pos(line))

File: day06/tuning_trouble.pl
#!/usr/bin/env perl
use v5.36;

sub marker_pos($buf, $len) {
    my @buf = split //, $buf;
    for my $i (0..$#buf-$len) {
        my %chars = map {$_ => 1} @buf[$i..$i+$len-1];
        return $i+$len if keys %chars == $len;
    }
    return -1;
}

while (<>) {
    say "Part 1: ", marker_pos($_, 4);
    say "Part 2: ", marker_pos($_, 14);
}
